File: python/runge_test.py
import math

from lagrange import lagrange
from lebesgue import lebesgue_const
import matplotlib.pyplot as plt
import numpy as np


# Runge example function
def runge_func(x):
    return 1.0/(1.0 + x**2)

# Returns a list of equidistant nodes over [a,b]
def equis_nodes(a, b, n):
    return np.linspace(a, b, n)

# Returns a list of n Chebyshev nodes over [a,b]
def cheby_nodes(a, b, n):
    # Chebyscev nodes over the interval [-1,1]
    l = [ math.cos((2*x+1)*math.pi/(2*n)) for x in range(0, n) ]
    # Affine transformation to the interval [a,b]
    return [ (a+b)/2.0 + (b-a)/2.0 * x for x in l ]

#
# Configurations
#
minpoints, maxpoints = 5, 17    # min and max number of nodes
ibeg, iend = -5, 5              # interval begin and end
step = 4
func = runge_func

#get_nodes = equis_nodes
get_nodes = cheby_nodes

xx = np.linspace(ibeg, iend, 100)
yy = [ func(x) for x in xx ]
for n in range(minpoints, maxpoints+1, step):
    _, ax = plt.subplots()
    ax.plot(xx, yy)
    xpoints = get_nodes(ibeg, iend, n)
    ypoints = [ func(x) for x in xpoints ]
    yy1 = lagrange(xpoints, ypoints, xx)
    ax.plot(xx, yy1)
    ax.scatter(xpoints, ypoints, s=15, color='black')
    ax.text(0.2, 0.9, 'n={}'.format(n),
            fontsize=20, transform=plt.gcf().transFigure)
    # Print Lebesgue constant
    lamb = lebesgue_const(xx, xpoints)
    print("n = {:2}   lambda = {}".format(n, lamb))

# Show the graphs
plt.show()

File: octave/runge_test.m
% TEST: interpolation of Runge's function test

% Runge's function definition
rungef = inline('1.0/(1.0 + x^2)');

%
% Configuration
%
minpoints = 3;  % min number of nodes
maxpoints = 20; % max number of nodes
step = 2;       % step between nodes
ibeg = -5;      % interval begin
iend =  5;      % interval end

% Returns a list of equidistant nodes over [a,b]
function x = equis_nodes(a, b, n)
    x = linspace(a, b, n);
end

% Returns a list of n Chebyshev nodes over [a,b]
function x = cheby_nodes(a, b, n)
    for i = 1:n
        % Chebyscev nodes over the interval [-1,1]
        x(i) = cos((2*(i-1)+1)*pi/(2*n));
        % Affine transformation to the interval [a,b]
        x(i) = (a+b)/2 + (b-a)/2 * x(i);
    end
end

% Get equidistant nodes function handle
%get_nodes = @equis_nodes;
get_nodes = @cheby_nodes;

xx = linspace(ibeg, iend, 100);
yy = arrayfun(rungef, xx);
for n = minpoints:step:maxpoints
    % Graph plot
    xpoints = get_nodes(ibeg, iend, n);
    ypoints = arrayfun(rungef, xpoints);
    yy1 = lagrange(xpoints, ypoints, xx);
    figure(n);
    plot(xx, yy, xx, yy1, xpoints, ypoints, 'o');
    % Print Lebesgue constant on the console
    lamb = lebesgue_const(xx, xpoints);
    printf('n=%d, labda=%f\n', n, lamb);
end

% Wait for input
pause;
